### ISSUE TITLE
Help optimizing D455 for outdoor agricultural mapping with RTAB-Map

### ISSUE THREAD

**Issue by VanjaLjubobratovic** [realsenseai/realsense-ros#2906]
Hello,
I'm working on a project which aims to use RealSense D455 and RTAB-Map in ROS2 in order to create
an areal view map of agricultural terrains. The idea is that those systems would already be in place on an autonomous plant sprayer, used for something else, and they would create the map on the side so you can monitor the plants.

I have already developed and set up everything needed for that, but the results I'm getting are not really satisfying so I was wondering if anyone could point me in the right direction.

Here are my results including my mosaic and the point cloud I produce:
[Mosaic](https://drive.google.com/file/d/19uqvot7Lusr1km17bDNnuPMj_si7u5Ip/view?usp=drive_link)
[PLY Cloud](https://drive.google.com/file/d/1OLBWOSGO9pXPQI9Gx-6y8T3mjJQa2WfZ/view?usp=drive_link)
[Screenshot of a part of the cloud in MeshLab](https://drive.google.com/file/d/11UNjKNPnTtnWPl1fZqN9OwGizo48JGU7/view?usp=drive_link)
[Image of the area so you have an idea what it actually looks like](https://drive.google.com/file/d/1GaTQffEx_rPsWLvt8AOboFNm69kMLT18/view?usp=drive_link)

Below you can see my rs_launch.py params in the form of a script I use for launching:
```
#!/bin/bash

resolution="1280x720x30"

LC_NUMERIC="en_US.UTF-8"; ros2 launch realsense2_camera rs_launch.py \
rgb_camera.profile:=$resolution \
depth_module.profile:=$resolution \
enable_gyro:=true \
enable_accel:=true \
unite_imu_method:=1 \
gyro_fps:=400 \
accel_fps:=200 \
enable_infra1:=false \
enable_infra2:=false \
enable_sync:=false \
align_depth.enable:=true \
pointcloud.enable:=false \
initial_reset:=true \
depth_module.enable_auto_exposure:=true \
decimation_filter.enable:=false \
temporal_filter.enable:=true \
spatial_filter.enable:=true \
```

Basically what I'm most interested in hearing is whether this looks like a camera problem, RTAB-Map problem or maybe both so I know what to look into.
Thanks!

**Comment by MartyG-RealSense** [comment]
Hi @VanjaLjubobratovic  Is the main problem the purple color tinting on the RGB image?  If it is then this is a known issue on the D455 camera model.  It is not a bug, but a characteristic of the D455 hardware that can occur when the RGB sensor is over-saturated with bright sunlight.

![image](https://github.com/IntelRealSense/realsense-ros/assets/41145062/e000fd42-5195-4675-b3a4-89e03704f38a)

Further information about this phenomenon can be found on the final page of the **Specification Clarification** document at the link below.

https://www.intelrealsense.com/download/13629/

D455 cameras manufactured after early 2022 had a design change to help reduce this effect.  Other actions that can be taken include setting a **Region of Interest** in the bottom half of the image, or applying an **ND filter** product over the top of the camera lenses on the outside of the camera.

**Comment by VanjaLjubobratovic** [comment]
@MartyG-RealSense, no, sorry, I was probably a bit unclear. 
My main problem is that I'm getting a lot of depth errors on the plants so they end up very fuzzy and smeared when you look from above (look at the garden area of the mosaic).

I realize it is hard to capture small and thin objects such as leaves in great detail, but I was wondering if, after all, there is something I can do to improve this or it just is what it is.

**Comment by MartyG-RealSense** [comment]
If it looks fuzzy and smeared then it may be more to do with the RGB color texture (which is being mapped onto the depth coordinates by align_depth) than it is to do with the depth data.  You may get a clearer image if you maximize the RGB **Sharpness** setting to a value of '100'.

**Comment by VanjaLjubobratovic** [comment]
Ok, I'll try that out and get back to you.

**Comment by VanjaLjubobratovic** [comment]
@MartyG-RealSense So maybe we misunderstood each other. What I meant by "smearing" and "fuzzy" will be illustrated by the following 2 images:
![image](https://github.com/IntelRealSense/realsense-ros/assets/34680384/af629bd5-9ede-4b11-a90e-780322e67864)
Have a look at those plants along the wall and on the stairs. They look fairly decent.
But when you look at this area of the cloud from the side or from above...
![image](https://github.com/IntelRealSense/realsense-ros/assets/34680384/0a45ca7e-5ea1-402c-9fef-23acd63bd1d8)

You can also see it on the columns and other objects
![image](https://github.com/IntelRealSense/realsense-ros/assets/34680384/a77cbc0b-5db4-4297-a724-6570ba78f980)

So basically what I wanted to ask is what mistake have I made to produce such results. Can I configure the camera better? Is it not a camera problem at all and I should try to change RTAB-Map parameters? Is it just my poor recording and I should try to capture everything from all the possible angles?

Thank you for your time!

**Comment by MartyG-RealSense** [comment]
The effect resembles a phenomenon called **point cloud spray** that is discussed at https://github.com/IntelRealSense/librealsense/issues/7021#issuecomment-670444359

Applying a **Visual Preset** pre-made camera configuration (also known as a **json file**) can help to deal with it.  https://github.com/IntelRealSense/realsense-ros/issues/2445 provides information about applying a preset in ROS.

**Comment by VanjaLjubobratovic** [comment]
Ok, I'll try that out. Thanks.

**Comment by VanjaLjubobratovic** [comment]
@MartyG-RealSense So applying the high accuracy preset helps somewhat, but the problem doesn't really go away. I have also been trying to mitigate it by recording with a greater angle towards the ground and that does also somewhat help since the errors aren't completely perpendicular to the Z axis now, but still they do persist on the objects that are captured on the edges of the field of view.

After reading various different discussions on the issue where no definitive answer is given, would I be correct in assuming there isn't much more I can do to fix this?

**Comment by MartyG-RealSense** [comment]
You could potentially position a second camera adjacent to the first one with their fields of view overlapping so that the overall combined field of view is wider and can see more of the scene.

Alternatively, you could try using 1280x720 resolution if you are not doing so already so that the image 'zooms out' a little and can include a little more information at the edges.

**Comment by VanjaLjubobratovic** [comment]
I will test 720p along with the preset since I have been using 480p since the test shown in the first comment because it seems to be producing fewer errors.
I will report on the results and after that we can probably close this issue.

Also, a question about the 2 camera suggestion: is this accomplished with rs_multi_camera_launch or do I need to do something else?

**Comment by MartyG-RealSense** [comment]
Intel have a ROS guide for "stitching" pointcloud views from two cameras together into a single pointcloud in ROS1, though the process would not be quite the same in ROS2.

https://github.com/IntelRealSense/realsense-ros/wiki/Showcase-of-using-2-cameras

**Comment by MartyG-RealSense** [comment]
Hi @VanjaLjubobratovic  Do you require further assistance with this case, please?  Thanks!

**Comment by VanjaLjubobratovic** [comment]
Hi, sorry I haven't been able to test it. Bad weather lately. Keep the issue open for a little while longer, please.

**Comment by MartyG-RealSense** [comment]
That's no problem at all to keep the issue open for a further time period.  Thanks very much for the update!

**Comment by VanjaLjubobratovic** [comment]
So after some experimentation I have managed to find a solution which I would call satisfactory.

In terms of camera settings:
- I've set **temporal filter off** as per documentation it can increase point cloud spray if the camera is moving
- I'm using the optimal profile which is **848x480x60** as per the documentation
- I'm using some of the camera presets. **High Accuracy** if there will be no filtering done on the cloud afterwards, and if you are filtering as I will be, maybe **medium density** can work great also, because it wont remove as much good points from your clouds.

In terms of camera operation:
- I'm recording at as big of an angle to the ground as I can since most of the errors are along the X axis, if the X axis points towards the ground, when I rotate the cloud to get an areal view, a lot of the spray will be covered up by good pixels

In terms of filtering:
- I have my custom node which is used to filter the point clouds produced by RTAB-Map and I'm using **distance thresholding** (0.5m to 4.0m). Already implemented by a RTAB-Map function for decoding messages on /rtabmap/mapData topic.
- My **custom filter** which calculates the median distance and standard deviation of all the points from the camera in the XY plane and removes those which don't fit into the specified interval. The effect is that point cloud edges are "trimmed". This removes a lot of outlier clusters where some outliers are being clustered together in the middle of empty space.
- **PCL Statistical Outlier Removal** filter which is specifically made to remove point cloud spray. It is used last because it is ineffective at removing clusters of outliers. Since I'm concerned with real time applications I was forced to make a multi threaded implementation which is decent for my needs.

The results of filtering on the point cloud **RECORDED WITH ORIGINAL CAMERA SETTINGS**:
BEFORE
![NoFilter](https://github.com/IntelRealSense/realsense-ros/assets/34680384/bc074d64-ac89-46ba-8c13-4f2e1b8107ad)

AFTER
![Filter](https://github.com/IntelRealSense/realsense-ros/assets/34680384/09f3e058-2d11-4cc9-a88a-571f98f8b017)

When all of the steps above are actually combined, the result should be quite good.
[Here's my code if someone's interested in implementation details](https://github.com/VanjaLjubobratovic/RTAB-Map-Orthophoto/blob/multi_cam/mosaicing_pkg/src/mosaicing_tools.cpp#L133C22-L133C33)

I hope someone finds this useful.
